thavalon-webapp/src/components/gameComponents/playerBoard.tsx

```diff
@@ -1,5 +1,177 @@
+        switch (message.messageType) {
+            case InboundMessageType.Snapshot:
+                const snapshot = message.data as Snapshot
+                handleGameMessage(snapshot.log[0]);
+                break;
+            case InboundMessageType.PlayerFocusChange:
+                const { displayName, visibility } = message.data as PlayerFocusChangeMessage;
+                playerFocusChanged(displayName, visibility);
+                break;
+            case InboundMessageType.GameMessage:
+                handleGameMessage(message.data as GameMessage);
+                break;
+        }
+    }
+
+    /**
+     * GameMessage specific message handler. This is needed because the GameMessage
+     * has internal types to parse.
+     * @param message The GameMessage from the server
+     */
+    function handleGameMessage(message: GameMessage): void {
+        switch (message.messageType) {
+            case GameMessageType.ProposalOrder:
+                setPlayerList(message.data as string[]);
+                break;
+        }
+    }
+
+    /**
+     * Updates the selected player list with a new name.
+     * @param name The player name to select.
+     */
+    function updateSelectedPlayers(name: string): void {
+        updateSet(selectedPlayers, name, setSelectedPlayers);
+    }
+
+    /**
+     * Sends a message to the server that the player is tabbed in or out.
+     */
+    function sendPlayerVisibilitychange(): void {
+        const connection = GameSocket.getInstance();
+        // visibilityState is either "hidden" or "visible." It can also be "prerender," but we don't talk about that.
+        const message = { messageType: OutboundMessageType.PlayerFocusChange, data: document.visibilityState };
+        connection.sendMessage(message);
+    }
+
+    /**
+     * Updates the tabbed out player list with a new player and visibility
+     * @param player The player whose focus has changed
+     * @param visibility The new visibility for that player
+     */
+    function playerFocusChanged(player: string, visibility: VisibilityState) {
+        const tempSet = new Set(tabbedOutPlayers.values());
+        if (visibility === "hidden") {
+            tempSet.add(player);
+        } else if (visibility === "visible") {
+            tempSet.delete(player);
+        }
+        setTabbedOutPlayers(tempSet);
+    }
+
+    /**
+     * Helper function to update a set.
+     * @param setToUpdate The set to update
+     * @param valueToToggle The value in the set to toggle
+     * @param reactSetter The React setter for the state to update
+     */
+    function updateSet<T>(setToUpdate: Set<T>, valueToToggle: T, reactSetter: React.Dispatch<React.SetStateAction<Set<T>>>): void {
+        const tempSet = new Set<T>(setToUpdate.values());
+        if (!tempSet.delete(valueToToggle)) {
+            tempSet.add(valueToToggle);
+        }
+        reactSetter(new Set<T>(tempSet.values()));
```

Comment:
Why does this make another copy instead of returning `tempSet`?

Reply:
It was test code. I've removed it.